fs: fix write methods param validation and docs (#41677)
nodejs/node: doc/api/fs.md

<!-- YAML
added: v10.0.0
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/41677
    description: The `data` parameter no longer accepts objects with an

>                  own `toString` function property.

Reviewer: In the end, did it ever accepted such objects? `node -e 'node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "string" } }))'` results in an abort on all release lines (tested on v14.x, v16.x, and v17.x). If this was never supported, maybe we can simply fix the doc and remove the added entry (I realize you added this because I asked for it, sorry for the confusion, this `fs.md` file is too big, I got confused).

                 buffers anymore.
-->

> * `buffer` {Buffer|TypedArray|DataView}

Reviewer: We probably need to add an entry to the changes list for this method as well.

Author: Yup, rebasing~

Reviewer: We'd still want an entry status for this `filehandle.write` I think https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L589-L597

Author: Ugh sorry, I've misread the signature.

This one is separated from `fh.write(string, ...)` in docs, so I consider string-related behaviours to be irrelevant here. The mentioned change is just a doc-only correction. I.e. it doesn't matter if first param is `string` or `string|Object`, because these params belong to "another method" anyway.

I think it would make more sense to keep the `v14.12.0 | The string parameter will stringify an object with an explicit toString function.` entry for `fh.write(string, ...)` method, but omit it in `fh.write(buffer, ...)` versions to avoid confusion.
Does it sound better?

Reviewer: Yes I think that would work – as long as the history doesn't imply that it's still supported, we should be good.

Reviewer: We also want to update these ones I think: 
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4442-L4450
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4469
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4526-L4529
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L4555
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L5746-L5749
https://github.com/nodejs/node/blob/f3b05ddf63d29e4f49ba405b25bc52d52279ee2f/doc/api/fs.md#L5767

Author: Eventually. Coercion in `fs.write`, `fs.writeFile` and `fs.writeFileSync` (as well as `fs.appendFile`, `fs.appendFileSync`) is doc-only deprecated ([DEP0162](https://github.com/nodejs/node/commit/4b198ffa7e681824dea10ed372a9ed7201c2c7f5)) atm; so these will be updated later.

Reviewer: If we want to make that a runtime deprecation for v18.x, we should hurry, the semver cutoff is next week.

Author: Affirmative. I can't promise anything right now, but will try to do it in time.

The kernel ignores the position argument and always appends the data to
the end of the file.

#### `filehandle.write(buffer, options)`

<!-- YAML
added: REPLACEME
-->

* `buffer` {Buffer|TypedArray|DataView}
* `options` {Object}
  * `offset` {integer} **Default:** `0`
  * `length` {integer} **Default:** `buffer.byteLength`
  * `position` {integer} **Default:** `null`
* Returns: {Promise}

Write `buffer` to the file.

Similar to the above `filehandle.write` function, this version takes an
optional `options` object. If no `options` object is specified, it will
default with the above values.

#### `filehandle.write(string[, position[, encoding]])`

<!-- YAML
added: v10.0.0
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/41677

>     description: The `string` parameter must be a primitive string.

Reviewer: ```suggestion
    description: The `string` parameter no longer accept objects with an
                 own `toString` property.
```

The kernel ignores the position argument and always appends the data to
the end of the file.

#### `filehandle.write(buffer, options)`

<!-- YAML
added: REPLACEME
-->

* `buffer` {Buffer|TypedArray|DataView}
* `options` {Object}
  * `offset` {integer} **Default:** `0`
  * `length` {integer} **Default:** `buffer.byteLength`
  * `position` {integer} **Default:** `null`
* Returns: {Promise}

Write `buffer` to the file.

Similar to the above `filehandle.write` function, this version takes an
optional `options` object. If no `options` object is specified, it will
default with the above values.

#### `filehandle.write(string[, position[, encoding]])`

<!-- YAML
added: v10.0.0
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/41677

>     description: The `string` parameter no longer accept objects with an

Reviewer: uhr my bad
```suggestion
    description: The `string` parameter no longer accepts objects with an
```

Author: Fixed for both methods, rebased on master

The kernel ignores the position argument and always appends the data to
the end of the file.

#### `filehandle.write(buffer, options)`

<!-- YAML
added: REPLACEME
-->

* `buffer` {Buffer|TypedArray|DataView}
* `options` {Object}
  * `offset` {integer} **Default:** `0`
  * `length` {integer} **Default:** `buffer.byteLength - offset`
  * `position` {integer} **Default:** `null`
* Returns: {Promise}

Write `buffer` to the file.

Similar to the above `filehandle.write` function, this version takes an
optional `options` object. If no `options` object is specified, it will
default with the above values.

#### `filehandle.write(string[, position[, encoding]])`

<!-- YAML
added: v10.0.0
changes:

>   - version: REPLACEME
>     pr-url: https://github.com/nodejs/node/pull/41677
>     description: The `string` parameter no longer accepts objects with an
>                  own `toString` function property.
>   - version: v14.12.0
>     pr-url: https://github.com/nodejs/node/pull/34993
>     description: The `string` parameter will stringify an object with an

Reviewer: In the end, did it ever accepted such objects? `node -e 'node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "string" } }))'` results in an abort on all release lines (tested on v14.x, v16.x, and v17.x). If this was never supported, maybe we can simply fix the doc and remove the added entry (I realize you added this because I asked for it, sorry for the confusion, this `fs.md` file is too big, I got confused).
In the end, did it ever accepted such objects? `node -e 'node -e 'fs.promises.open("/dev/null").then(fh => fh.write({ toString() { return "string" } }))'` results in an abort on all release lines (tested on v14.x, v16.x, and v17.x). If this was never supported, maybe we can simply fix the doc and remove the added entry (I realize you added this because I asked for it, sorry for the confusion, this `fs.md` file is too big, I got confused).

```suggestion
```

Author: Yep, all affected methods weren't working with coercion in the first place. Annihilated all related entries.
I initially assumed that "rewriting history" in the docs would be a bad idea; but yes, it is much cleaner this way, and it won't mislead people who use older versions of Node.

> this `fs.md` file is too big

It sure is. 😄 I hope, something like removal of whole Callback API in foreseeable future will reduce it, and both code and docs will become less confusing.

    description: Passing an invalid callback to the `callback` argument
                 now throws `ERR_INVALID_ARG_TYPE` instead of
                 `ERR_INVALID_CALLBACK`.
  - version: v14.12.0
    pr-url: https://github.com/nodejs/node/pull/34993
    description: The `buffer` parameter will stringify an object with an

>                  explicit `toString` function.

Reviewer: Hum it looks to me that this does work though, and is still supported 🤔 

```console
$ node -e 'fs.write(1, { toString() { return "string\n" } }, Function.prototype)'
string
```

Reviewer: Arf, I'm still falling for it, I forgot that we document that method twice with different signature, nevermind.

-->

* `fd` {integer}

> * `buffer` {Buffer|TypedArray|DataView}

Reviewer: We'd need a change entry here too: https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L4354-L4367

Author: Resolved in the same way as [filehandle.write](https://github.com/nodejs/node/pull/41677#discussion_r841283211): redundant changelog entry is removed, string-related changes are kept in `fs.write(fd, string, ...)` method.

Note: string version of this method *does* coerce Objects even after this PR. This is doc-only segregation.

-->

* `fd` {integer}

> * `buffer` {Buffer|TypedArray|DataView}

Reviewer: Here as well https://github.com/nodejs/node/blob/ae1f6db0886a540dcf33f27bb4f13c37e0c8899e/doc/api/fs.md#L5806-L5814

Author: Resolved in the same way as [filehandle.write](https://github.com/nodejs/node/pull/41677#discussion_r841283211): redundant changelog entry is removed, string-related changes were already added for `fs.writeSync(fd, string, ...)` version.
